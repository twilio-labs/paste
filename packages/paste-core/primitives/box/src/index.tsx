import styled from '@emotion/styled';
import {compose, layout, space, background, border, boxShadow, position, flexbox, system} from 'styled-system';
import css from '@styled-system/css';
import {
  LayoutProps,
  SpaceProps,
  BackgroundProps,
  BorderProps,
  ShadowProps,
  PositionProps,
  FlexboxProps,
  TypographyProps,
} from '@twilio-paste/style-props';
import {
  CursorProperty,
  AppearanceProperty,
  AnimationProperty,
  TransformProperty,
  TransformOriginProperty,
  VisibilityProperty,
  WhiteSpaceProperty,
  UserSelectProperty,
  PointerEventsProperty,
  BoxSizingProperty,
  ResizeProperty,
  TransitionProperty,
  ListStyleTypeProperty,
  ListStylePositionProperty,
  ListStyleImageProperty,
  ObjectFitProperty,
  ObjectPositionProperty,
  BackgroundAttachmentProperty,
  OutlineProperty,
  FloatProperty,
  WillChangeProperty,
} from 'csstype';
import {PseudoPropStyles} from './PseudoPropStyles';
import {BoxPropTypes} from './BoxPropTypes';

export interface BaseBoxProps
  extends React.HTMLAttributes<any>,
    LayoutProps,
    SpaceProps,
    BackgroundProps,
    BorderProps,
    ShadowProps,
    PositionProps,
    FlexboxProps {
  as?: keyof JSX.IntrinsicElements;
  content?: string;
  cursor?: CursorProperty;
  appearance?: AppearanceProperty;
  animation?: AnimationProperty;
  transform?: TransformProperty;
  transformOrigin?: TransformOriginProperty<string>;
  visibility?: VisibilityProperty;
  whiteSpace?: WhiteSpaceProperty;
  userSelect?: UserSelectProperty;
  pointerEvents?: PointerEventsProperty;
  boxSizing?: BoxSizingProperty;
  resize?: ResizeProperty;
  transition?: TransitionProperty;
  listStyleType?: ListStyleTypeProperty;
  listStylePosition?: ListStylePositionProperty;
  listStyleImage?: ListStyleImageProperty;
  objectFit?: ObjectFitProperty;
  objectPosition?: ObjectPositionProperty<string>;
  backgroundAttachment?: BackgroundAttachmentProperty;
  outline?: OutlineProperty<string>;
  float?: FloatProperty;
  willChange?: WillChangeProperty;
  textDecoration?: TypographyProps['textDecoration'];
  /** Typed as any because Box can literally be any HTML element */
  ref?: any | null;
}

interface PseudoStylesProps {
  _after?: BaseBoxProps;
  _before?: BaseBoxProps;
  _focus?: BaseBoxProps;
  _hover?: BaseBoxProps;
  _active?: BaseBoxProps;
  _pressed?: BaseBoxProps;
  _selected?: BaseBoxProps;
  _focusWithin?: BaseBoxProps;
  _invalid?: BaseBoxProps;
  _disabled?: BaseBoxProps;
  _grabbed?: BaseBoxProps;
  _expanded?: BaseBoxProps;
  _checked?: BaseBoxProps;
  _mixed?: BaseBoxProps;
  _odd?: BaseBoxProps;
  _even?: BaseBoxProps;
  _visited?: BaseBoxProps;
  _readOnly?: BaseBoxProps;
  _first?: BaseBoxProps;
  _last?: BaseBoxProps;
  _groupHover?: BaseBoxProps;
  _notFirst?: BaseBoxProps;
  _notLast?: BaseBoxProps;
  _placeholder?: BaseBoxProps;
}

export interface BoxProps extends BaseBoxProps, PseudoStylesProps {}

const extraConfig = system({
  backgroundColor: {
    property: 'backgroundColor',
    scale: 'backgroundColors',
  },
  borderColor: {
    property: 'borderColor',
    scale: 'borderColors',
  },
  animation: true,
  appearance: true,
  transform: true,
  transformOrigin: true,
  visibility: true,
  whiteSpace: true,
  userSelect: true,
  pointerEvents: true,
  boxSizing: true,
  cursor: true,
  resize: true,
  transition: true,
  listStyleType: true,
  listStylePosition: true,
  listStyleImage: true,
  objectFit: true,
  objectPosition: true,
  backgroundAttachment: {
    property: 'backgroundAttachment',
  },
  outline: true,
  float: true,
  willChange: true,
  textDecoration: true,
});

const getPseudoStyles = (props: BoxProps): {} => {
  const pseudoProps = Object.keys(props).filter(propName => propName.startsWith('_'));

  if (pseudoProps.length === 0) {
    return {};
  }

  const pseudoStyles = {};
  pseudoProps.forEach(pseudoProp => {
    if (PseudoPropStyles[pseudoProp] != null) {
      pseudoStyles[PseudoPropStyles[pseudoProp]] = props[pseudoProp];
    }
  });

  return css(pseudoStyles);
};

/* eslint-disable emotion/syntax-preference */
export const Box = styled.div(
  {
    boxSizing: 'border-box',
  },
  compose(
    space,
    layout,
    flexbox,
    background,
    border,
    boxShadow,
    position,
    extraConfig
  ),
  getPseudoStyles
) as React.FC<BoxProps>;
/* eslint-enable */

Box.displayName = 'Box';

if (process.env.NODE_ENV === 'development') {
  Box.propTypes = BoxPropTypes;
}

export * from './SafelySpreadProps';
