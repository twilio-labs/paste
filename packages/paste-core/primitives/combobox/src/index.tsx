import Downshift from '@twilio-paste/dropdown-library';

export {Downshift as ComboboxPrimitive};

export {useCombobox as useComboboxPrimitive} from '@twilio-paste/dropdown-library';

export type {
  UseComboboxInterface as UseComboboxPrimitiveInterface,
  UseComboboxProps as UseComboboxPrimitiveProps,
  UseComboboxState as UseComboboxPrimitiveState,
  UseComboboxReturnValue as UseComboboxPrimitiveReturnValue,
  UseComboboxStateChange as UseComboboxPrimitiveStateChange,
} from '@twilio-paste/dropdown-library';
