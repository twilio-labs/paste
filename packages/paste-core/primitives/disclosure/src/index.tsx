// https://reakit.io/docs/disclosure/
export {
  useDisclosureState as useDisclosurePrimitiveState,
  Disclosure as DisclosurePrimitive,
  DisclosureContent as DisclosurePrimitiveContent,
} from '@twilio-paste/reakit-library';
