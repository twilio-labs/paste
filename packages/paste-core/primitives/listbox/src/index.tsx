export * from "./ListboxPrimitive";
export * from "./ListboxPrimitiveItem";
export * from "./useListboxPrimitiveState";

export { CompositeGroup as ListboxPrimitiveGroup } from "@twilio-paste/reakit-library";
export type { CompositeGroupProps as ListboxPrimitiveGroupProps } from "@twilio-paste/reakit-library";
