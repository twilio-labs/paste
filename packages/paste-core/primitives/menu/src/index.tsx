export {
  useMenuState as useMenuPrimitiveState,
  Menu as MenuPrimitive,
  MenuGroup as MenuPrimitiveGroup,
  MenuItem as MenuPrimitiveItem,
  MenuItemCheckbox as MenuPrimitiveItemCheckbox,
  MenuItemRadio as MenuPrimitiveItemRadioMenuItemRadio,
  MenuButton as MenuPrimitiveButton,
  MenuSeparator as MenuPrimitiveSeparator,
} from '@twilio-paste/reakit-library';

export type {
  MenuProps as MenuPrimitiveProps,
  MenuGroupProps as MenuPrimitiveGroupProps,
  MenuItemProps as MenuPrimitiveItemProps,
  MenuItemCheckboxProps as MenuPrimitiveItemCheckboxProps,
  MenuItemRadioProps as MenuPrimitiveItemRadioMenuItemRadioProps,
  MenuButtonProps as MenuPrimitiveButtonProps,
  MenuSeparatorProps as MenuPrimitiveSeparatorProps,
} from '@twilio-paste/reakit-library';
