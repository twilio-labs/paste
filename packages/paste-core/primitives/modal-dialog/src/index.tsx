/**
 * If you’re wondering why we wrapped that package into our own, we reasoned that it
 * would be best for our consumers’ DX, with reasons such as:
 *
 * - If we want to migrate the underlying nuts and bolts in the future, Twilio products
 *   that depend on this primitive would need to replace all occurrences of
 *   `import … from ‘@reach/dialog’` to `import … from ‘@some-new/package’`.
 *   By wrapping it in `@twilio-paste/modal-dialog-primitive`, this refactor can be avoided.
 *   The only change would be a version bump in the package.json file.
 * - We can more strictly enforce semver and backwards compatibility than some of
 *   our dependencies.
 * - We can control when to provide an update and which versions we whitelist, to
 *   help reduce potential bugs our consumers may face.
 * - We can control which APIs we expose. For example, we may chose to enable or
 *   disable usage of certain undocumented APIs.
 */
import {DialogContent as ModalDialogContent, DialogOverlay as ModalDialogOverlay} from '@reach/dialog';

export {ModalDialogContent, ModalDialogOverlay};
