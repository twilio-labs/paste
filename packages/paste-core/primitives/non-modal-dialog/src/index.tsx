export {
  usePopoverState as useNonModalDialogPrimitiveState,
  PopoverState as NonModalDialogPrimitiveState,
  PopoverInitialState as NonModalDialogPrimitivePopoverInitialState,
  Popover as NonModalDialogPrimitive,
  PopoverProps as NonModalDialogPrimitiveProps,
  PopoverDisclosure as NonModalDialogDisclosurePrimitive,
  PopoverDisclosureProps as NonModalDialogDisclosurePrimitiveProps,
  PopoverArrow as NonModalDialogArrowPrimitive,
  PopoverArrowProps as NonModalDialogArrowPrimitiveProps,
} from '@twilio-paste/reakit-library';
