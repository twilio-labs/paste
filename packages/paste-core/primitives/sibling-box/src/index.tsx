export { SiblingBox } from "./SiblingBox";
export type { SiblingBoxProps } from "./SiblingBox";
