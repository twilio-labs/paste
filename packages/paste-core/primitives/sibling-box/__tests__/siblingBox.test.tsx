import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Theme} from '@twilio-paste/theme';
import {SiblingBox} from '../src';

describe('SiblingBox render', () => {
  it('it should render', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider theme="console">
          <SiblingBox
            width="sizeSquare50"
            _hover={{
              backgroundColor: 'colorBackgroundBody',
            }}
            _hoverSibling={{
              paddingLeft: 'space50',
            }}
            _activeSibling={{
              color: 'colorText',
            }}
            _invalidSibling={{
              backgroundColor: 'colorBackground',
            }}
            _disabledSibling={{
              borderColor: 'colorBorder',
            }}
            _focusSibling={{
              margin: 'space10',
            }}
            _checkedSibling={{
              padding: 'space10',
            }}
            _boxChild={{
              marginBottom: 'space30',
            }}
            _checkedSiblingAndBoxChild={{
              paddingBottom: 'space30',
            }}
            _checkedAndDisabledSibling={{
              width: 'sizeIcon50',
            }}
            _checkedAndActiveSibling={{
              height: 'sizeSquare110',
            }}
            _checkedAndFocusSibling={{
              minHeight: 'size110',
            }}
            _checkedAndInvalidSibling={{
              marginRight: 'space80',
            }}
            _checkedAndHoverSibling={{
              borderRadius: 'borderRadius10',
            }}
          >
            child
          </SiblingBox>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
