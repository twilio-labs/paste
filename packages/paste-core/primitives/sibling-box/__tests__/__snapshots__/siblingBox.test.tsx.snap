// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiblingBox render it should render 1`] = `
.emotion-1 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-0 {
  box-sizing: border-box;
  width: 1rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 120ms;
  transition: all 120ms;
}

.emotion-0:hover {
  background-color: rgb(255,255,255);
}

input[type=checkbox]:active + label .emotion-0 {
  color: rgb(40,42,43);
}

input[type=checkbox]:checked + label .emotion-0,
input[type=checkbox][aria-checked=mixed] + label .emotion-0 {
  padding: 0.125rem;
}

input[type=checkbox]:checked:active + label .emotion-0,
input[type=checkbox][aria-checked=mixed]:active + label .emotion-0 {
  height: 2.5rem;
}

input[type=checkbox]:checked + label .emotion-0 > *,
input[type=checkbox][aria-checked=mixed] + label .emotion-0 > * {
  padding-bottom: 0.5rem;
}

input[type=checkbox]:checked:disabled + label .emotion-0,
input[type=checkbox][aria-checked=mixed]:disabled + label .emotion-0 {
  width: 1.75rem;
}

input[type=checkbox]:checked:focus + label .emotion-0,
input[type=checkbox][aria-checked=mixed]:focus + label .emotion-0 {
  min-height: 70.5rem;
}

input[type=checkbox]:checked:hover:not(:disabled) + label .emotion-0,
input[type=checkbox][aria-checked=mixed]:hover:not(:disabled) + label .emotion-0 {
  border-radius: 2px;
}

input[type=checkbox][aria-invalid=true]:checked + label .emotion-0 {
  margin-right: 1.75rem;
}

input[type=checkbox] + label .emotion-0 > * {
  margin-bottom: 0.5rem;
}

input[type=checkbox]:disabled + label .emotion-0 {
  border-color: rgb(136,140,142);
}

input[type=checkbox]:focus + label .emotion-0 {
  margin: 0.125rem;
}

input[type=checkbox]:hover:not(:disabled):not(:checked) + label .emotion-0 {
  padding-left: 1rem;
}

input[type=checkbox][aria-invalid=true] + label .emotion-0 {
  background-color: rgb(248,248,249);
}

<div
  className="emotion-1"
>
  <div
    aria-hidden="true"
    className="emotion-0"
    display="inline-flex"
    width="sizeSquare50"
  >
    child
  </div>
</div>
`;
